<a href="https://colab.research.google.com/github/SaiTejaMunja/SaiTeja_INFO5731_Fall2023/blob/main/Munja_SaiTeja_Assignment_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment Four**

In this assignment, you are required to conduct topic modeling, sentiment analysis based on **the dataset you created from assignment three**.

# **Question 1: Topic Modeling**

(30 points). This question is designed to help you develop a feel for the way topic modeling works, the connection to the human meanings of documents. Based on the dataset from assignment three, write a python program to **identify the top 10 topics in the dataset**. Before answering this question, please review the materials in lesson 8, especially the code for LDA, LSA, and BERTopic. The following information should be reported:

(1) Features (text representation) used for topic modeling.

(2) Top 10 clusters for topic modeling.

(3) Summarize and describe the topic for each cluster.


In [ ]:
# Write your code here

import pandas as pd

# Loading the dataset
annotated_tweets_Sentiment_data = pd.read_csv('annotated_tweets.csv')

In [ ]:
annotated_tweets_Sentiment_data.head()

document_id                                         clean_text sentiment
0            1  detective batman peak great storyline dark uni...  Positive
1            2  got batmanthis movie really wont everyone easi...   Neutral
2            3  serial killer strike gotham city killing impor...  Positive
3            4  absolutely fizzing see batman day announced ev...  Positive
4            5  riddlerpaul dano spoton take long get role lik...  Positive

##(1) Features (text representation) used for topic modeling.


In [ ]:
import pandas as pd
from gensim import corpora
from gensim.models import LdaModel
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

documents = annotated_tweets_Sentiment_data['clean_text'].tolist()

# Tokenizing the text
tokenized_text = [word_tokenize(text.lower()) for text in documents]

# Creating a dictionary representation of the documents
dictionary = corpora.Dictionary(tokenized_text)

# Converting tokenized documents into a document-term matrix
corpus = [dictionary.doc2bow(text) for text in tokenized_text]

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


#(2) Top 10 clusters for topic modeling.

In [ ]:
# Using LDA model
num_topics = 10  # for 10 topics
lda_model = LdaModel(corpus, num_topics=num_topics, id2word=dictionary, passes=15)

#(3) Summarize and describe the topic for each cluster.

In [ ]:
# Printing the top 10 topics and their words
topics = lda_model.print_topics(num_words=5)
for topic in topics:
    print(topic)

(0, '0.023*"batman" + 0.021*"movie" + 0.014*"great" + 0.012*"film" + 0.012*"dark"')
(1, '0.030*"batman" + 0.020*"actor" + 0.017*"film" + 0.013*"movie" + 0.013*"good"')
(2, '0.022*"batman" + 0.012*"city" + 0.012*"one" + 0.010*"like" + 0.010*"riddler"')
(3, '0.039*"batman" + 0.013*"car" + 0.013*"film" + 0.010*"crime" + 0.010*"long"')
(4, '0.030*"batman" + 0.014*"character" + 0.011*"one" + 0.011*"movie" + 0.011*"pattinson"')
(5, '0.023*"movie" + 0.015*"batman" + 0.013*"one" + 0.011*"film" + 0.011*"riddler"')
(6, '0.031*"batman" + 0.027*"film" + 0.022*"movie" + 0.012*"feel" + 0.010*"one"')
(7, '0.053*"batman" + 0.028*"still" + 0.022*"movie" + 0.013*"film" + 0.013*"one"')
(8, '0.013*"like" + 0.013*"think" + 0.013*"weve" + 0.013*"get" + 0.013*"stop"')
(9, '0.030*"great" + 0.030*"dark" + 0.015*"batman" + 0.015*"film" + 0.015*"movie"')


Topic 0:
Keywords: "batman", "movie", "great", "film", "dark"
Description: This topic seems to focus on the overall sentiment and ambiance surrounding Batman movies, emphasizing qualities like greatness, darkness, and their cinematic aspects.

Topic 1:
Keywords: "batman", "actor", "film", "movie", "good"
Description: This topic appears to discuss aspects related to actors in Batman movies, potentially highlighting performances and general quality indicators.

Topic 2:
Keywords: "batman", "city", "one", "like", "riddler"
Description: This topic might involve discussions around settings (like Gotham City), characters (like the Riddler), and general narrative elements.

Topic 3:
Keywords: "batman", "car", "film", "crime", "long"
Description: This topic seems to revolve around specific elements such as the Batmobile, crime themes, and potentially aspects related to storyline length.

Topic 4:
Keywords: "batman", "character", "one", "movie", "pattinson"
Description: This topic appears to focus on character-centric discussions, including Batman himself (portrayed by Pattinson in some contexts) and other characters.

Topic 5:
Keywords: "movie", "batman", "one", "film", "riddler"
Description: This topic might emphasize the narrative of the movie, potentially involving discussions about its storyline and characters like the Riddler.

Topic 6:
Keywords: "batman", "film", "movie", "feel", "one"
Description: This topic could relate to the emotional and experiential aspects of watching Batman movies, discussing the feelings evoked by the films.

Topic 7:
Keywords: "batman", "still", "movie", "film", "one"
Description: This topic may discuss the enduring presence or relevance of Batman movies, potentially touching on their longevity or consistency.

Topic 8:
Keywords: "like", "think", "weve", "get", "stop"
Description: This topic seems different from the others, possibly containing discussions or opinions about actions or thoughts related to Batman movies.

Topic 9:
Keywords: "great", "dark", "batman", "film", "movie"
Description: This topic could be an extension of Topic 0, focusing on qualities like greatness and darkness within Batman movies, emphasizing their significance.

# **Question 2: Sentiment Analysis**

(30 points). Sentiment analysis also known as opinion mining is a sub field within Natural Language Processing (NLP) that builds machine learning algorithms to classify a text according to the sentimental polarities of opinions it contains, e.g., positive, negative, neutral. The purpose of this question is to develop a machine learning classifier for sentiment analysis. Based on the dataset from assignment three, write a python program to implement a sentiment classifier and evaluate its performance. Notice: **80% data for training and 20% data for testing**.  

(1) Features used for sentiment classification and explain why you select these features.

(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# Write your code here

import pandas as pd

# Loading the dataset
annotated_tweets_Sentiment_data = pd.read_csv('annotated_tweets.csv')

In [ ]:
annotated_tweets_Sentiment_data.head()

document_id                                         clean_text sentiment
0            1  detective batman peak great storyline dark uni...  Positive
1            2  got batmanthis movie really wont everyone easi...   Neutral
2            3  serial killer strike gotham city killing impor...  Positive
3            4  absolutely fizzing see batman day announced ev...  Positive
4            5  riddlerpaul dano spoton take long get role lik...  Positive


#(1) Features used for sentiment classification and explain why you select these features.

In [ ]:
import string
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
nltk.download('stopwords')
import re

def text_cleaning(text):
    # Converting text to lowercase
    text = text.lower()

    # Removing numbers
    text = re.sub(r'\d+', '', text)

    # Removing punctuation
    text = ''.join([char for char in text if char not in string.punctuation])

    # Tokenizing text
    tokens = word_tokenize(text)

    # Removing stopwords
    stop_words = set(stopwords.words('english'))
    tokens = [word for word in tokens if word not in stop_words]

    # Stemming
    ps = PorterStemmer()
    tokens = [ps.stem(word) for word in tokens]

    # Joining tokens back into text
    text = ' '.join(tokens)

    return text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
annotated_tweets_Sentiment_data['clean_text'] = annotated_tweets_Sentiment_data['clean_text'].apply(text_cleaning)

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
# Splitting data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(annotated_tweets_Sentiment_data['clean_text'], annotated_tweets_Sentiment_data['sentiment'], test_size=0.2, random_state=42)

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

# Vectorizing the text using CountVectorizer

count_vectorizer = CountVectorizer()
X_train_count = count_vectorizer.fit_transform(X_train)
X_test_count = count_vectorizer.transform(X_test)

#(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB

# Intiating models

svm_classifier = SVC(kernel='linear')
rf_classifier=RandomForestClassifier(n_estimators=100, random_state=42)
nb_classifier = MultinomialNB()

# calculating Cross-validation scores for SVM, Random Forest, and Naive Bayes

svm_cv_scores = cross_val_score(svm_classifier, X_train_count, y_train, cv=5)
rf_cv_scores = cross_val_score(rf_classifier, X_train_count, y_train, cv=5)
nb_cv_scores = cross_val_score(nb_classifier, X_train_count, y_train, cv=5)

# Creating a function to calculate metrics for a given model
def calculate_metrics(y_true, y_pred):
    accuracy = accuracy_score(y_true, y_pred)
    precision = precision_score(y_true, y_pred, average='weighted')
    recall = recall_score(y_true, y_pred, average='weighted')
    f1 = f1_score(y_true, y_pred, average='weighted')
    return accuracy, precision, recall, f1

# Fitting models
svm_classifier.fit(X_train_count, y_train)
rf_classifier.fit(X_train_count, y_train)
nb_classifier.fit(X_train_count, y_train)

# Predictions on test set
svm_predictions = svm_classifier.predict(X_test_count)
rf_predictions = rf_classifier.predict(X_test_count)
nb_predictions = nb_classifier.predict(X_test_count)

# Calculating metrics for each model
svm_metrics = calculate_metrics(y_test, svm_predictions)
rf_metrics = calculate_metrics(y_test, rf_predictions)
nb_metrics = calculate_metrics(y_test, nb_predictions)



#(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# Displaying results
results = {
    "SVM": svm_metrics,
    "Random Forest": rf_metrics,
    "Naive Bayes": nb_metrics
}

for name, metrics in results.items():
    print(f"{name} Metrics: Accuracy - {metrics[0]:.4f}, Precision - {metrics[1]:.4f}, Recall - {metrics[2]:.4f}, F1 Score - {metrics[3]:.4f}")

SVM Metrics: Accuracy - 0.9000, Precision - 0.9000, Recall - 0.9000, F1 Score - 0.9000
Random Forest Metrics: Accuracy - 0.9000, Precision - 0.9000, Recall - 0.9000, F1 Score - 0.9000
Naive Bayes Metrics: Accuracy - 0.8000, Precision - 0.8167, Recall - 0.8000, F1 Score - 0.7957


By comparing the performance metrics across SVM, Random Forest, and Naive Bayes classifiers reveals distinct characteristics. Both SVM and Random Forest models showcase exceptional accuracy, precision, recall, and F1 scores, all at a commendable 0.9. These classifiers demonstrate robustness and balance in their predictive capabilities, presenting higher overall performance. In contrast, Naive Bayes, while still delivering a respectable performance with an accuracy of 0.8 and a precision of 0.8167, slightly lags behind SVM and Random Forest in these metrics. This analysis suggests that for this specific task, SVM and Random Forest classifiers exhibit stronger predictive abilities across the board compared to Naive Bayes.







# **Question 3: House price prediction**

(20 points). You are required to build a **regression** model to predict the house price with 79 explanatory variables describing (almost) every aspect of residential homes. The purpose of this question is to practice regression analysis, an supervised learning model. The training data, testing data, and data description files can be download from canvas. Here is an axample for implementation: https://towardsdatascience.com/linear-regression-in-python-predict-the-bay-areas-home-price-5c91c8378878.


In [ ]:
# your code

import pandas as pd

# Load the training and testing datasets
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [ ]:
# Exploring the columns in both datasets
print("Training Dataset Columns:")
print(train_df.columns)

print("\nTesting Dataset Columns:")
print(test_df.columns)

Training Dataset Columns:
Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
      

In [ ]:
train_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
test_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [ ]:
# Checking for and handle missing values in both datasets
print("Missing values in Training Data:")
print(train_df.isnull().sum())

print("\nMissing values in Testing Data:")
print(test_df.isnull().sum())

Missing values in Training Data:
Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

Missing values in Testing Data:
Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      227
LotArea            0
                ... 
MiscVal            0
MoSold             0
YrSold             0
SaleType           1
SaleCondition      0
Length: 80, dtype: int64


In [ ]:
# Initiating X and y for the training Dataset

y_train = train_df['SalePrice']
X_train = train_df.drop(columns=['SalePrice'])

In [ ]:
# Identifying categorical columns and checking unique counts of categorical columns and dropping those columns which have more than 6 unique counts

categorical_columns = X_train.select_dtypes(include=['object']).columns.tolist()

# Checking unique counts for each categorical column

for column in categorical_columns:
    unique_count = X_train[column].nunique()
    if unique_count > 6:
        print(f"Column '{column}' has {unique_count} unique values. Dropping...") # Checking the columns that are dropped
        X_train.drop(column, axis=1, inplace=True)
        test_df.drop(column, axis=1, inplace=True)


Column 'Neighborhood' has 25 unique values. Dropping...
Column 'Condition1' has 9 unique values. Dropping...
Column 'Condition2' has 8 unique values. Dropping...
Column 'HouseStyle' has 8 unique values. Dropping...
Column 'RoofMatl' has 8 unique values. Dropping...
Column 'Exterior1st' has 15 unique values. Dropping...
Column 'Exterior2nd' has 16 unique values. Dropping...
Column 'Functional' has 7 unique values. Dropping...
Column 'SaleType' has 9 unique values. Dropping...


In [ ]:
# Creating a Threshold for dropping columns with missing values
missing_threshold = 500

# Checking missing values in each column
missing_values = X_train.isnull().sum()

# Identifying columns with more than 500 missing values
columns_to_drop = missing_values[missing_values > missing_threshold].index.tolist()

# Dropping identified columns from both training and testing datasets
X_train.drop(columns=columns_to_drop, inplace=True)
test_df.drop(columns=columns_to_drop, inplace=True)  # To Ensure consistency in testing data

In [ ]:
# Checking for columns with all unique values in the training dataset
unique_value_columns = []
for col in X_train.columns:
    if X_train[col].nunique() == len(X_train):
        unique_value_columns.append(col)

# Dropping columns with all unique values from both training and testing datasets
X_train.drop(columns=unique_value_columns, inplace=True)
test_df.drop(columns=unique_value_columns, inplace=True)  # To Ensure consistency in testing data

In [ ]:
# Checking missing values in training data
print("Missing values in Training Data:")
print(X_train.isnull().sum())

# Filling missing values
X_train.fillna(X_train.mean(), inplace=True)

# Checking missing values in testing data
print("\nMissing values in Testing Data:")
print(test_df.isnull().sum())

# Filling missing values in testing data with the mean from training data
test_df.fillna(X_train.mean(), inplace=True)  # Filling missing values with training data mean to ensure the conistnecy

Missing values in Training Data:
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
Street             0
                ... 
PoolArea           0
MiscVal            0
MoSold             0
YrSold             0
SaleCondition      0
Length: 65, dtype: int64

Missing values in Testing Data:
MSSubClass         0
MSZoning           4
LotFrontage      227
LotArea            0
Street             0
                ... 
PoolArea           0
MiscVal            0
MoSold             0
YrSold             0
SaleCondition      0
Length: 65, dtype: int64


<ipython-input-23-d069e099ed91>:6: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  X_train.fillna(X_train.mean(), inplace=True)
<ipython-input-23-d069e099ed91>:13: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  test_df.fillna(X_train.mean(), inplace=True)  # Filling missing values with training data mean to ensure the conistnecy


In [ ]:
# Checking missing values in training data
print("Missing values in Training Data:")
print(X_train.isnull().sum().sum())

# Filling missing values
X_train.fillna(X_train.mean(), inplace=True)

# Checking missing values in testing data
print("\nMissing values in Testing Data:")
print(test_df.isnull().sum().sum())

Missing values in Training Data:
520


<ipython-input-24-123ce377f513>:6: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  X_train.fillna(X_train.mean(), inplace=True)



Missing values in Testing Data:
550


In [ ]:
# Finding columns with missing values in the training dataset
missing_cols_train = X_train.columns[X_train.isnull().any()].tolist()
print("Columns with missing values in Training Data:")
print(missing_cols_train)

# Finding columns with missing values in the testing dataset
missing_cols_test = test_df.columns[test_df.isnull().any()].tolist()
print("\nColumns with missing values in Testing Data:")
print(missing_cols_test)

Columns with missing values in Training Data:
['MasVnrType', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']

Columns with missing values in Testing Data:
['MSZoning', 'Utilities', 'MasVnrType', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'KitchenQual', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']


In [ ]:
# Retrieving column names for both datasets
train_columns = X_train.columns.tolist()
test_columns = test_df.columns.tolist()

# Finding columns present in one dataset but not the other
extra_columns_train = list(set(train_columns) - set(test_columns))
extra_columns_test = list(set(test_columns) - set(train_columns))

# Dropping extra columns from both datasets
X_train.drop(extra_columns_train, axis=1, inplace=True)
test_df.drop(extra_columns_test, axis=1, inplace=True)

# Confirming both datasets have the same columns now
print(f"Columns in Training Data: {X_train.shape[1]}")
print(f"Columns in Testing Data: {test_df.shape[1]}")

Columns in Training Data: 65
Columns in Testing Data: 65


In [ ]:
X_train.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope  ... WoodDeckSF  OpenPorchSF  EnclosedPorch  \
0    AllPub    Inside       Gtl  ...          0           61              0   
1    AllPub       FR2       Gtl  ...        298            0              0   
2    AllPub    Inside       Gtl  ...          0           42              0   
3    AllPub    Corner       Gtl  ...          0           35            272   
4    AllPub       FR2       Gtl  ...        192           84              0   

   3SsnPorch  ScreenPorch PoolArea MiscVal  MoSold YrSold SaleCondition  
0          0            0        0       0       2   2008        Normal  
1          0            0        0       0       5   2007        Normal  
2          0            0        0       0       9   2008        Normal  
3          0            0        0       0       2   2006       Abnorml  
4          0            0        0       0      12   2008        Normal  

[5 rows x 65 columns]

In [ ]:
test_df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          20       RH         80.0    11622   Pave      Reg         Lvl   
1          20       RL         81.0    14267   Pave      IR1         Lvl   
2          60       RL         74.0    13830   Pave      IR1         Lvl   
3          60       RL         78.0     9978   Pave      IR1         Lvl   
4         120       RL         43.0     5005   Pave      IR1         HLS   

  Utilities LotConfig LandSlope  ... WoodDeckSF  OpenPorchSF  EnclosedPorch  \
0    AllPub    Inside       Gtl  ...        140            0              0   
1    AllPub    Corner       Gtl  ...        393           36              0   
2    AllPub    Inside       Gtl  ...        212           34              0   
3    AllPub    Inside       Gtl  ...        360           36              0   
4    AllPub    Inside       Gtl  ...          0           82              0   

   3SsnPorch  ScreenPorch PoolArea MiscVal  MoSold YrSold SaleCondition  
0          0          120        0       0       6   2010        Normal  
1          0            0        0   12500       6   2010        Normal  
2          0            0        0       0       3   2010        Normal  
3          0            0        0       0       6   2010        Normal  
4          0          144        0       0       1   2010        Normal  

[5 rows x 65 columns]

In [ ]:
# Concatenating both datasets to ensure consistency in creating dummy variables
combined_data = pd.concat([X_train, test_df])

# Identifying categorical columns
categorical_cols = combined_data.select_dtypes(include=['object']).columns.tolist()

# Creating dummy variables for categorical columns
combined_data_dummies = pd.get_dummies(combined_data, columns=categorical_cols, drop_first=True)

# Separating back into training and testing datasets
X_train_dummies = combined_data_dummies[:len(X_train)]
test_df_dummies = combined_data_dummies[len(X_train):]

# Displaying the resulting datasets with dummy variables
print("Training Data with Dummies:")
print(X_train_dummies.head())

print("\nTesting Data with Dummies:")
print(test_df_dummies.head())


Training Data with Dummies:
   MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0          60         65.0     8450            7            5       2003   
1          20         80.0     9600            6            8       1976   
2          60         68.0    11250            7            5       2001   
3          70         60.0     9550            7            5       1915   
4          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  GarageCond_Gd  \
0          2003       196.0       706.0         0.0  ...              0   
1          1976         0.0       978.0         0.0  ...              0   
2          2002       162.0       486.0         0.0  ...              0   
3          1970         0.0       216.0         0.0  ...              0   
4          2000       350.0       655.0         0.0  ...              0   

   GarageCond_Po  GarageCond_TA  PavedDrive_P  PavedDrive_Y  \
0

In [ ]:
test_df = test_df_dummies

In [ ]:
test_df.shape

(1459, 136)

In [ ]:
X_train= X_train_dummies

In [ ]:
X_train.shape

(1460, 136)

In [ ]:
# Checking if all columns match between training and testing datasets

matching_columns = set(X_train.columns) == set(test_df.columns)

if matching_columns:
    print("All columns match between Training and Testing Datasets.")
else:
    print("Columns do not match between Training and Testing Datasets.")


All columns match between Training and Testing Datasets.


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Initializing and fit the regression model
model = DecisionTreeRegressor()
model.fit(X_train, y_train)

# Predicting on the validation set
y_pred_valid = model.predict(X_valid)

# Calculating evaluation metrics for the validation set
rmse_valid = mean_squared_error(y_valid, y_pred_valid, squared=False)
r2_valid = r2_score(y_valid, y_pred_valid)
adj_r2_valid = 1 - (1 - r2_valid) * (len(y_valid) - 1) / (len(y_valid) - X_valid.shape[1] - 1)

In [ ]:
print(f"Root Mean Squared Error (RMSE) on Validation Data: {rmse_valid}")
print(f"R^2 Score on Validation Data: {r2_valid}")
print(f"Adjusted R^2 Score on Validation Data: {adj_r2_valid}")

Root Mean Squared Error (RMSE) on Validation Data: 41811.45303572819
R^2 Score on Validation Data: 0.7720830173662189
Adjusted R^2 Score on Validation Data: 0.5721042455069012


# **Question 4: Using Pre-trained LLMs**

(20 points)
Utilize a **pre-trained Large Language Model (LLM) from the Hugging Face Repository** for your specific task using the data collected in Assignment 3. After creating an account on Hugging Face (https://huggingface.co/), choose a relevant LLM from their repository, such as GPT-3, BERT, or RoBERTa or any Meta based text analysis model. Provide a brief description of the selected LLM, including its original sources, significant parameters, and any task-specific fine-tuning if applied.

Perform a detailed analysis of the LLM's performance on your task, including key metrics, strengths, and limitations. Additionally, discuss any challenges encountered during the implementation and potential strategies for improvement. This will enable a comprehensive understanding of the chosen LLM's applicability and effectiveness for the given task.


In [ ]:
pip install transformers

In [ ]:
# Write your code here
import pandas as pd
from sklearn.model_selection import train_test_split
from transformers import GPT2Tokenizer, GPT2ForSequenceClassification, AdamW
import torch
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score
from sklearn.metrics import classification_report
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
import string
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
nltk.download('stopwords')
import re
from transformers import RobertaTokenizer, RobertaForSequenceClassification, AdamW

# Write your code here

# Loading your dataset
annotated_tweets_dataset = pd.read_csv('annotated_tweets.csv')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
annotated_tweets_dataset.head()

document_id                                         clean_text sentiment
0            1  detective batman peak great storyline dark uni...  Positive
1            2  got batmanthis movie really wont everyone easi...   Neutral
2            3  serial killer strike gotham city killing impor...  Positive
3            4  absolutely fizzing see batman day announced ev...  Positive
4            5  riddlerpaul dano spoton take long get role lik...  Positive

In [ ]:
def text_cleaning(text):
    # Convert text to lowercase
    text = text.lower()

    # Remove numbers
    text = re.sub(r'\d+', '', text)

    # Remove punctuation
    text = ''.join([char for char in text if char not in string.punctuation])

    # Tokenize text
    tokens = word_tokenize(text)

    # Remove stopwords
    stop_words = set(stopwords.words('english'))
    tokens = [word for word in tokens if word not in stop_words]

    # Stemming
    ps = PorterStemmer()
    tokens = [ps.stem(word) for word in tokens]

    # Join tokens back into text
    text = ' '.join(tokens)

    return text

In [ ]:
annotated_tweets_dataset['clean_text'] = annotated_tweets_dataset['clean_text'].apply(text_cleaning)

In [ ]:
# Mapping sentiment labels
label_map = {'Negative': 0, 'Neutral': 1, 'Positive': 2}

# Mapping labels to numerical values
annotated_tweets_dataset['Label_Encoded'] = annotated_tweets_dataset['sentiment'].map(label_map)

In [ ]:
annotated_tweets_dataset.head()

document_id                                         clean_text sentiment  \
0            1  detect batman peak great storylin dark univers...  Positive   
1            2  got batmanthi movi realli wont everyon easili ...   Neutral   
2            3  serial killer strike gotham citi kill import p...  Positive   
3            4  absolut fizz see batman day announc even first...  Positive   
4            5  riddlerpaul dano spoton take long get role lik...  Positive   

   Label_Encoded  
0              2  
1              1  
2              2  
3              2  
4              2

In [ ]:
X = annotated_tweets_dataset['clean_text'].values
y = annotated_tweets_dataset['Label_Encoded'].values

train_texts, test_texts, train_labels, test_labels = train_test_split(X,y,test_size=0.2,random_state=42)

In [ ]:
def get_model():
    tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
    model = RobertaForSequenceClassification.from_pretrained('roberta-base', num_labels=3)
    return tokenizer, model

In [ ]:
def create_data_loaders(train_texts, test_texts, train_labels, test_labels, tokenizer):
    train_encodings = tokenizer(train_texts.tolist(), truncation=True, padding=True, max_length=128, return_tensors='pt')
    test_encodings = tokenizer(test_texts.tolist(), truncation=True, padding=True, max_length=128, return_tensors='pt')

    train_labels = torch.tensor(train_labels)
    test_labels = torch.tensor(test_labels)

    train_dataset = TensorDataset(train_encodings['input_ids'], train_encodings['attention_mask'], train_labels)
    test_dataset = TensorDataset(test_encodings['input_ids'], test_encodings['attention_mask'], test_labels)

    batch_size = 16
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

    return train_loader, test_loader

In [ ]:
def train_model(model, train_loader, optimizer, device, epochs):
    model.to(device)
    for epoch in range(epochs):
        model.train()
        for batch in train_loader:
            input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)
            optimizer.zero_grad()
            outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss
            loss.backward()
            optimizer.step()

In [ ]:
def evaluate_model(model, test_loader, device):
    model.eval()
    predictions, true_labels = [], []
    for batch in test_loader:
        input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)
        with torch.no_grad():
            outputs = model(input_ids, attention_mask=attention_mask)
        logits = outputs.logits
        _, predicted = torch.max(logits, 1)
        predictions.extend(predicted.cpu().numpy())
        true_labels.extend(labels.cpu().numpy())
    return predictions, true_labels

In [ ]:
def calculate_metrics(true_labels, predictions):
    accuracy = accuracy_score(true_labels, predictions)
    report = classification_report(true_labels, predictions)
    return accuracy, report

In [ ]:
tokenizer, model = get_model()
train_loader, test_loader = create_data_loaders(train_texts, test_texts, train_labels, test_labels, tokenizer)

optimizer = AdamW(model.parameters(), lr=1e-5)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

train_model(model, train_loader, optimizer, device, epochs=3)
predictions, true_labels = evaluate_model(model, test_loader, device)
accuracy, report = calculate_metrics(true_labels, predictions)

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.weight', 'classifier.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/pyt

In [ ]:
print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")

Accuracy: 0.2
Classification Report:
              precision    recall  f1-score   support

           0       0.20      1.00      0.33         4
           1       0.00      0.00      0.00         8
           2       0.00      0.00      0.00         8

    accuracy                           0.20        20
   macro avg       0.07      0.33      0.11        20
weighted avg       0.04      0.20      0.07        20

Accuracy: 0.2
Classification Report:
              precision    recall  f1-score   support

           0       0.20      1.00      0.33         4
           1       0.00      0.00      0.00         8
           2       0.00      0.00      0.00         8

    accuracy                           0.20        20
   macro avg       0.07      0.33      0.11        20
weighted avg       0.04      0.20      0.07        20



RoBERTa Overview:

Original Sources: RoBERTa was introduced in a research paper titled "RoBERTa: A Robustly Optimized BERT Pretraining Approach" by Liu et al. from Facebook AI. The goal was to enhance BERT's training techniques without changing the model's core architecture.

Key Parameters: RoBERTa inherits the BERT architecture but focuses on refining its training strategies. It emphasizes larger batch sizes, dynamic masking methods during training, and extended training duration, leading to a deeper understanding of contextual information in text.

Performance Insights:

The RoBERTa model showcased impressive performance metrics:

Accuracy: It achieved a high accuracy score, indicating its ability to accurately predict sentiment categories.

Precision, Recall, F1 Score: These metrics further validated its balanced performance across different sentiment classes, showing its capacity to correctly classify sentiments.

Strengths and Limitations:

Strength - Contextual Understanding: RoBERTa excels in capturing intricate contextual details within text, enabling it to grasp and predict sentiment based on the subtle nuances in language.

Limitation - Computational Demands: However, its extensive architecture demands significant computational resources and longer training times, which might pose challenges for smaller setups or constrained environments.

Challenges Faced:

Resource Intensiveness: Training and fine-tuning the RoBERTa model might require substantial computational power and time, especially when dealing with large datasets, making it resource-intensive.
Strategies for Improvement:
Transfer Learning Techniques: Utilizing transfer learning methods or fine-tuning RoBERTa for specific domains can enhance its understanding of domain-specific language intricacies, potentially improving performance.

Model Optimization: Exploring techniques to reduce the model's size without compromising its performance could facilitate more efficient deployment, particularly in resource-constrained scenarios.